README.md
# Node-Red SonosPollyTTS

[![NPM version][npm-version-image]][npm-url]
[![NPM downloads per month][npm-downloads-month-image]][npm-url]
[![NPM downloads total][npm-downloads-total-image]][npm-url]
[![MIT License][license-image]][license-url]
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
[![Donate via PayPal](https://img.shields.io/badge/Donate-PayPal-blue.svg?style=flat-square)](https://www.paypal.me/techtoday) 
[![Facebook][facebook-image]][facebook-url]


# BIG UPDATE ANNOUCEMENT
On 24 December 2020 a new big update has been released with the new node [TTS-Ultimate](https://github.com/Supergiovane/node-red-contrib-tts-ultimate).<br/>
Now, all nodes developed by Supergiovane, does have the "-ultimate" suffix, so you can find it and rely on sturdy and well developed nodes.<br/>
TTS-Ultimate supports now both Amazon Polly and Google TTS!<br/>
Check it out and install TTS-Ultimate instead of SonosPollyTTS

# UPDATE PATH FROM SONOSPOLLYTTS TO TTS-ULTIMATE
Supergiovane takes care about your brain and your time.<br/>
Install TTS-Ultimate. Both SonosPollyTTS and TTS-Ultimate can cohexist.<br/>
Then just delete your old SonosPollyTTS nodes and replace it with TTS-Ultimate nodes.<br/>
The cache will remain the same. Your own audio files and hailing files won't be touched. You'll find it again in TTS-Ultimate<br/>

# DEPRECATION WARNING
On 24 December 2020 this node has been deprecated and superseeded by [TTS-Ultimate](https://github.com/Supergiovane/node-red-contrib-tts-ultimate).<br/>
SonosPollyTTS will not be developed anymore.


    
[license-image]: https://img.shields.io/badge/license-MIT-blue.svg
[license-url]: https://github.com/Supergiovane/node-red-contrib-sonospollytts/master/LICENSE
[npm-url]: https://npmjs.org/package/node-red-contrib-sonospollytts
[npm-version-image]: https://img.shields.io/npm/v/node-red-contrib-sonospollytts.svg
[npm-downloads-month-image]: https://img.shields.io/npm/dm/node-red-contrib-sonospollytts.svg
[npm-downloads-total-image]: https://img.shields.io/npm/dt/node-red-contrib-sonospollytts.svg
[facebook-image]: https://img.shields.io/badge/Visit%20me-Facebook-blue
[facebook-url]: https://www.facebook.com/supergiovaneDev

CHANGELOG.md
![Sample Node](img/logo.png) 

[![Donate via PayPal](https://img.shields.io/badge/Donate-PayPal-blue.svg?style=flat-square)](https://www.paypal.me/techtoday) 

<br/>

# THIS NODE HAS BEEN DEPRECATED. PLEASE INSTALL TTS-ULTIMATE INSTEAD
[TTS-Ultimate](https://www.npmjs.com/package/node-red-contrib-tts-ultimate)  
[Update path instructions](https://github.com/Supergiovane/node-red-contrib-sonospollytts) 
<br/>
<br/>

<br/>
<br/>

<p>
<b>Version 2.0.5</b> December 2020<br/>
- FIX: if you pass a numeric value as payload, the node thows an error in the status and the payload is not handled.</br>
- FIX: if you disable the Hailing, you could hear "zero" instead of not having Haliling at all.
</p>
<p>
<b>Version 2.0.4</b> December 2020<br/>
- NEW: multiple sonospollytts node coordinator. Now, if you use more than one sonospollytts node, the server will coordinate the play between all, without conflicts.<br/>
</p>
<p>
<b>Version 2.0.3</b> December 2020<br/>
- FIX: sonospollytts can remain stuck on "Busy handling queue.." if if fails grouping more than one speaker. Fixed.<br/>
</p>
<p>
<b>Version 2.0.2</b> December 2020<br/>
- FIX: if you hear "zero" instead of hailing, you need to re-select again the hailing in the config node. This is due to an old setting. Currently this has been work-arounded, so you should hear the hailing instead of "zero", but please select again the hailing, because this workaround is time limited.<br/>
- NEW: now, when sonos becames disconnected or the ethernet connection is down, the node won't accept any new messages until the reconnetion is re-established. The queue of messages is flushed as well.<br/>
- Stability improvements.<br/>
</p>
<p>
<b>Version 2.0.1</b> December 2020<br/>
- NEW: automatic resume of radio station. If the station support seeking, the node'll resume at exct time position.<br/>
</p>
<p>
<b>Version 2.0.0</b> December 2020<br/>
- NEW: Completely rewritten the code, to getting rid of very old implementations.<br/>
- Resume queued music now works correctly everytime.<br/>
- BREAKING CHANGE: Node v.10.0.0 or newer is needed.<br/>
</p>
<p>
<b>Version 1.1.39</b> December 2020<br/>
- NEW: resume music queue after TTS speech. Once finished playing the voice speak, the music queue restart at the exact position, at the exact track time.<br/>
- BUGFIX: fixing some issue when inbound msg to the node are very frequent. Should resume the music correctly.<br/>
</p>
<p>
<b>Version 1.1.37</b> December 2020<br/>
- NEW: resume music queue after TTS speech. Once finished playing the voice speak, the music queue restart at the exact position, at the exact track time.<br/>
</p>
<p>
<b>Version 1.1.36</b> November 2020<br/>
- Whenever node-red is restarted or you make a deploy while Sonos is playing music, the node won't stop Sonos players anymore.<br/>
- Speeded up the closing process.<br/>
</p>
<p>
<b>Version 1.1.35</b> October 2020<br/>
- HOTFIX: Fix possible polly voice discovery issues if running in docker or under homeassistant.<br/>
- Added a warning to restart node-red after each change in the config window.<br/>
</p>
<p>
<b>Version 1.1.34</b> October 2020<br/>
- HOTFIX: Fix possible sonos discovery issue<br/>
</p>
<p>
<b>Version 1.1.33</b> October 2020<br/>
- FIX: fixed problem preventing OwnFile node to accept files in non standard node-red installations (dockerized images, as homeassistant plugin etc..). Thanks @koburg for raising the issue.<br/>
- Now, whenever you upload a new file in Ownfile or new Hailing in Sonospollytts node, the new file is autimatically selected.<br/>
</p>
<p>
<b>Version 1.1.32</b> October 2020<br/>
- Optimized the code to better handling HTTP calls for UI list.<br/>
- NEW: from now, the avaiable voice list is downloaded directly from AWS Polly website. Non more "forgotten" voices.<br/>
- Reordering of voices by language, to better find your preferred one.<br/>
</p>
<p>
<b>Version 1.1.31</b> October 2020<br/>
- FIX: fixed possible issue in saving/retrieving own audio files and own audio hailings, for node-red running in non standard way, like docker, home assistant plugin etc...<br/>
</p>
<p>
<b>Version 1.1.30</b> September 2020<br/>
- If you run node-red behind natted docker (for example kubertness) and you've added sonos devices to the group list, the devices belonging to the group doesn't accept volume setting. This has been fixed. Thanks @JorinL for raising the issue.<br/>
- NEW: Added more voices and removed old ones no more present in Polly.<br/>
</p>
<p>
<b>Version 1.1.28</b> September 2020<br/>
- Updated sonos API to 1.14.0<br/>
- Fixed an issue where if you power cycle your sonos and it comes up too fast, the node won't detect the disconnection and won't reset the link to sonos.<br/>
- Major code revision, to clean up old things.<br/>
</p>
<p>
<b>Version 1.1.27</b> August 2020<br/>
- Fixed an odd issue. Now "msg.nohailing=true" temporarly (and not permanently anymore) disables the hailing. If you send a new payload without msg.nohailing=true, the hailing will be heard again (that's the standard and intended behaviour.)<br/>
- Sometime the node remains in "Processing" state. Added a fix to that.
</p>
<p>
<b>Version 1.1.26</b> Juli 2020<br/>
- Stability improvement whenever internet connection is broken while downloading TTS audio from Amazon AWS service.<br/>
- Full support for SSML syntax (https://docs.aws.amazon.com/polly/latest/dg/supportedtags.html).<br/>
</p>
<p>
<b>Version 1.1.25</b> June 2020<br/>
- Update sonos api to 1.13.0<br/>
- Fix some minor glitches in new added nodes.<br/>
</p>
<p>
<b>Version 1.1.23</b> March 2020 - In the middle of Coronavirus emergency in Italy<br/>
- Help and README enhancement with new samples<br/>
- Fix some minor glitches in new added nodes.<br/>
</p>
<p>
<b>Version 1.1.22</b> March 2020 - In the middle of Coronavirus emergency in Italy<br/>
- Enhancement: Automatic discover for new node, now automatically adds the first avaiable player.<br/>

sonospollytts/ownfile.js



module.exports = function (RED) {
    var formidable = require('formidable');
    var fs = require('fs');
    var path = require('path');

    function ownfile(config) {
        RED.nodes.createNode(this, config);
        var node = this;
        node.selectedFile = config.selectedFile || "";
        node.userDir = path.join(RED.settings.userDir , "sonospollyttsstorage"); // 09/03/2020 Storage of sonospollytts
        
        // Receive new files from html
        RED.httpAdmin.post("/node-red-contrib-sonospollytts", function (req, res) {
            var form = new formidable.IncomingForm();
            form.parse(req, function (err, fields, files) {
                if (err) { };
                // Allow only mp3
                if (files.customTTS.name.indexOf(".mp3") !== -1) {
                    var newPath = path.join(node.userDir , "ttspermanentfiles", "OwnFile_" + files.customTTS.name);
                    fs.rename(files.customTTS.path, newPath, function (err) { });
                }
            });
            res.json({ status: 220 });
            res.end;
        });



        // 27/02/2020 Get list of filenames starting with OwnFile_
        RED.httpAdmin.get("/getOwnFilesList", RED.auth.needsPermission('ownfile.read'), function (req, res) {
            var jListOwnFiles = [];
            var sName = "";
            try {
                fs.readdirSync(path.join(node.userDir , "ttspermanentfiles")).forEach(file => {
                    if (file.indexOf("OwnFile_") > -1) {
                        sName = file.replace("OwnFile_", '').replace(".mp3", '');
                        jListOwnFiles.push({ name: sName, filename: file });
                    }
                });

            } catch (error) { }
            res.json(jListOwnFiles)
        });

        // 27/02/2020 Delete OwnFile_
        RED.httpAdmin.get("/deleteOwnFile", RED.auth.needsPermission('ownfile.read'), function (req, res) {
            try {
                if (req.query.FileName == "DELETEallFiles") {
                    // Delete all OwnFiles_
                    try {
                        fs.readdir(path.join(node.userDir , "ttspermanentfiles"), (err, files) => {
                            files.forEach(function (file) {
                                if (file.indexOf("OwnFile_") !== -1) {
                                    RED.log.warn("SonospollyTTS: Deleted file " + path.join(node.userDir , "ttspermanentfiles" , file));
                                    try {
                                        fs.unlink(path.join(node.userDir , "ttspermanentfiles" , file)), err => { };
                                    } catch (error) { }
                                }
                            });
                        });

                    } catch (error) { }
                } else {
                    // Delete only one file
                    try {
                        var newPath = path.join(node.userDir , "ttspermanentfiles" , req.query.FileName);
                        try {
                            fs.unlinkSync(newPath)
                        } catch (error) { }

                    } catch (error) { }
                }
            } catch (err) {
            }
            res.json({ status: 220 });
        });


        this.on('input', function (msg) {
            if (msg.hasOwnProperty("selectedFile")) {
                if (msg.hasOwnProperty("selectedFile")) msg.payload = "OwnFile_" + msg.selectedFile.replace(".mp3", "") + ".mp3";
                node.send(msg);
            } else {
                if (msg.payload !== undefined && msg.payload === true || msg.payload === false) {
                    msg.payload = node.selectedFile;
                    node.send(msg);
                }
            }
        });
    }
    RED.nodes.registerType("ownfile", ownfile);
};

sonospollytts/sonospollytts-config.js
module.exports = function (RED) {
    'use strict';

    var AWS = require('aws-sdk');
    var fs = require('fs');
    var path = require("path");
    var formidable = require('formidable');
    const oOS = require('os');
    const sonos = require('sonos');

    AWS.config.update({
        region: 'us-east-1'
    });

    // Configuration Node Register
    function PollyConfigNode(config) {
        RED.nodes.createNode(this, config);
        var node = this;
        node.noderedipaddress = typeof config.noderedipaddress === "undefined" ? "" : config.noderedipaddress;
        var params = {
            accessKeyId: node.credentials.accessKey,
            secretAccessKey: node.credentials.secretKey,
            apiVersion: '2016-06-10'
        };
        node.polly = new AWS.Polly(params);
        node.userDir = path.join(RED.settings.userDir, "sonospollyttsstorage"); // 09/03/2020 Storage of sonospollytts (otherwise, at each upgrade to a newer version, the node path is wiped out and recreated, loosing all custom files)
        node.whoIsUsingTheServer = ""; // Client node.id using the server, because only a sonospollytts node can use the serve at once.

        // 03/06/2019 you can select the temp dir
        //#region "SETUP PATHS"
        // 26/10/2020 Check for path and create it if doens't exists
        function setupDirectory(_aPath) {

            if (!fs.existsSync(_aPath)) {
                // Create the path
                try {
                    fs.mkdirSync(_aPath);
                    return true;
                } catch (error) { return false; }
            } else {
                return true;
            }
        }
        if (!setupDirectory(node.userDir)) {
            RED.log.error('SonosPollyTTS: Unable to set up MAIN directory: ' + node.userDir);
        }
        if (!setupDirectory(path.join(node.userDir, "ttsfiles"))) {
            RED.log.error('SonosPollyTTS: Unable to set up cache directory: ' + path.join(node.userDir, "ttsfiles"));
        } else {
            RED.log.info('SonosPollyTTS: TTS cache set to ' + path.join(node.userDir, "ttsfiles"));
        }
        if (!setupDirectory(path.join(node.userDir, "hailingpermanentfiles"))) {
            RED.log.error('SonosPollyTTS: Unable to set up hailing directory: ' + path.join(node.userDir, "hailingpermanentfiles"));
        } else {
            RED.log.info('SonosPollyTTS: hailing path set to ' + path.join(node.userDir, "hailingpermanentfiles"));
            // 09/03/2020 Copy defaults to the userDir
            fs.readdirSync(path.join(__dirname, "hailingpermanentfiles")).forEach(file => {
                try {
                    fs.copyFileSync(path.join(__dirname, "hailingpermanentfiles", file), path.join(node.userDir, "hailingpermanentfiles", file));
                } catch (error) { }
            });
        }
        if (!setupDirectory(path.join(node.userDir, "ttspermanentfiles"))) {
            RED.log.error('SonosPollyTTS: Unable to set up permanent files directory: ' + path.join(node.userDir, "ttspermanentfiles"));
        } else {
            RED.log.info('SonosPollyTTS: permanent files path set to ' + path.join(node.userDir, "ttspermanentfiles"));
            // 09/03/2020 // Copy the samples of permanent files into the userDir
            fs.readdirSync(path.join(__dirname, "ttspermanentfiles")).forEach(file => {
                try {
                    fs.copyFileSync(path.join(__dirname, "ttspermanentfiles", file), path.join(node.userDir, "ttspermanentfiles", file));
                } catch (error) { }
            });
        }
        //#endregion

        //#region SONOSPOLLY NODE
        // ######################################################
        // 21/03/2019 Endpoint for retrieving the default IP
        RED.httpAdmin.get("/sonospollyTTSGetEthAddress", RED.auth.needsPermission('PollyConfigNode.read'), function (req, res) {
            var oiFaces = oOS.networkInterfaces();
            var jListInterfaces = [];
            try {
                Object.keys(oiFaces).forEach(ifname => {
                    // Interface with single IP
                    if (Object.keys(oiFaces[ifname]).length === 1) {
                        if (Object.keys(oiFaces[ifname])[0].internal == false) jListInterfaces.push({ name: ifname, address: Object.keys(oiFaces[ifname])[0].address });
                    } else {
                        var sAddresses = "";
                        oiFaces[ifname].forEach(function (iface) {
                            if (iface.internal == false && iface.family === "IPv4") sAddresses = iface.address;
                        });
                        if (sAddresses !== "") jListInterfaces.push({ name: ifname, address: sAddresses });
                    }
                })
            } catch (error) { }
            if (jListInterfaces.length > 0) {
                res.json(jListInterfaces[0].address); // Retunr the first usable IP
            } else {
                res.json("NO ETH INTERFACE FOUND");
            }

        });

        // 20/03/2020 in the middle of coronavirus, get the sonos groups
        RED.httpAdmin.get("/sonosgetAllGroups", RED.auth.needsPermission('PollyConfigNode.read'), function (req, res) {
            var jListGroups = [];
            try {
                const discovery = new sonos.AsyncDeviceDiscovery()
                discovery.discover().then((device, model) => {
                    return device.getAllGroups().then((groups) => {
                        //RED.log.warn('Groups ' + JSON.stringify(groups, null, 2))
                        for (let index = 0; index < groups.length; index++) {
                            const element = groups[index];
                            jListGroups.push({ name: element.Name, host: element.host })
                        }
                        res.json(jListGroups)
                        //return groups[0].CoordinatorDevice().togglePlayback()
                    })
                }).catch(e => {
                    RED.log.warn('SonosPollyTTS: Error in discovery ' + e);
                    res.json("ERRORDISCOVERY");
                })
            } catch (error) { }
        });


        // 09/03/2020 Get list of filenames in hailing folder
        RED.httpAdmin.get("/getHailingFilesList", RED.auth.needsPermission('PollyConfigNode.read'), function (req, res) {
            var jListOwnFiles = [];
            var sName = "";
            try {
                fs.readdirSync(path.join(node.userDir, "hailingpermanentfiles")).forEach(file => {
                    if (file.indexOf("Hailing_") > -1) {
                        sName = file.replace("Hailing_", "").replace(".mp3", "");
                        jListOwnFiles.push({ name: sName, filename: file });
                    }
                });

            } catch (error) { }
            res.json(jListOwnFiles)
        });

        // 09/03/2020 Delete Hailing
        RED.httpAdmin.get("/deleteHailingFile", RED.auth.needsPermission('PollyConfigNode.read'), function (req, res) {
            // Delete the file
            try {
                var newPath = path.join(node.userDir, "hailingpermanentfiles", req.query.FileName);
                fs.unlinkSync(newPath)
            } catch (error) { }
            res.json({ status: 220 });
        });

        // 09/03/2020 Receive new hailing files from html
        RED.httpAdmin.post("/sonospollyttsHailing", function (req, res) {
            var form = new formidable.IncomingForm();
            form.parse(req, function (err, fields, files) {
                if (err) { };
                // Allow only mp3
                if (files.customHailing.name.indexOf(".mp3") !== -1) {
                    var newPath = path.join(node.userDir, "hailingpermanentfiles", "Hailing_" + files.customHailing.name);
                    fs.rename(files.customHailing.path, newPath, function (err) { });
                }
            });
            res.json({ status: 220 });
            res.end;
        });

        // 26/10/2020 Supergiovane, get the updated voice list from AWS. 
        RED.httpAdmin.get("/pollygetvoices", RED.auth.needsPermission('PollyConfigNode.read'), function (req, res) {
            // node.refreshVoices().then(function (resolve) {
            //     res.json(resolve);
            // }).catch(function (reject) { 
            //     res.json(reject);
            //  });
            
            var jListVoices = [];
            try {
                // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Polly.html#describeVoices-property
                var jfiltroVoci = {
                    //Engine: standard | neural,
                    //IncludeAdditionalLanguageCodes: true 
                    //LanguageCode: arb | cmn-CN | cy-GB | da-DK | de-DE | en-AU | en-GB | en-GB-WLS | en-IN | en-US | es-ES | es-MX | es-US | fr-CA | fr-FR | is-IS | it-IT | ja-JP | hi-IN | ko-KR | nb-NO | nl-NL | pl-PL | pt-BR | pt-PT | ro-RO | ru-RU | sv-SE | tr-TR,
                    //NextToken: "STRING_VALUE"
                };
                node.polly.describeVoices(jfiltroVoci, function (err, data) {
                    if (err) {
                        RED.log.warn('SonosPollyTTS: Error getting polly voices ' + err);
                        jListVoices.push({ name: "Error retrieving voices. Check your AWS credentials and restart node-red", id: "Ivy" })
                        res.json(jListVoices)
                    } else {
                        for (let index = 0; index < data.Voices.length; index++) {
                            const oVoice = data.Voices[index];
                            jListVoices.push({ name: oVoice.LanguageName + " (" + oVoice.LanguageCode + ") " + oVoice.Name + " - " + oVoice.Gender, id: oVoice.Id })
                        }
                        res.json(jListVoices)
                    }
                });

            } catch (error) {
                jListVoices.push({ name: "Error " + error, id: "Ivy" })
                res.json(jListVoices)
            }
        });

        // ########################################################
        //#endregion



        //#region OWNFILE NODE
        // ######################################################

        // Receive new files from html
        RED.httpAdmin.post("/sonospollyttsOwnFile", function (req, res) {
            var form = new formidable.IncomingForm();
            form.parse(req, function (err, fields, files) {
                if (err) { };
                // Allow only mp3
                if (files.customTTS.name.indexOf(".mp3") !== -1) {
                    var newPath = path.join(node.userDir, "ttspermanentfiles", "OwnFile_" + files.customTTS.name);
                    fs.rename(files.customTTS.path, newPath, function (err) { });
                }
            });
            res.json({ status: 220 });
            res.end;
        });

        // 27/02/2020 Get list of filenames starting with OwnFile_
        RED.httpAdmin.get("/getOwnFilesList", RED.auth.needsPermission('PollyConfigNode.read'), function (req, res) {
            var jListOwnFiles = [];
            var sName = "";
            try {
                fs.readdirSync(path.join(node.userDir, "ttspermanentfiles")).forEach(file => {
                    if (file.indexOf("OwnFile_") > -1) {
                        sName = file.replace("OwnFile_", '').replace(".mp3", '');
                        jListOwnFiles.push({ name: sName, filename: file });
                    }
                });

            } catch (error) { }
            res.json(jListOwnFiles)
        });

        // 27/02/2020 Delete OwnFile_
        RED.httpAdmin.get("/deleteOwnFile", RED.auth.needsPermission('PollyConfigNode.read'), function (req, res) {
            try {
                if (req.query.FileName == "DELETEallFiles") {
                    // Delete all OwnFiles_
                    try {
                        fs.readdir(path.join(node.userDir, "ttspermanentfiles"), (err, files) => {
                            files.forEach(function (file) {
                                if (file.indexOf("OwnFile_") !== -1) {
                                    RED.log.warn("SonospollyTTS: Deleted file " + path.join(node.userDir, "ttspermanentfiles", file));
                                    try {
                                        fs.unlinkSync(path.join(node.userDir, "ttspermanentfiles", file));
                                    } catch (error) { }
                                }
                            });
                        });

                    } catch (error) { }
                } else {
                    // Delete only one file
                    try {
                        var newPath = path.join(node.userDir, "ttspermanentfiles", req.query.FileName);
                        try {
                            fs.unlinkSync(newPath)
                        } catch (error) { }

                    } catch (error) { }
                }
            } catch (err) {
            }
            res.json({ status: 220 });
        });
        // ########################################################
        //#endregion

        node.oWebserver; // 11/11/2019 Stores the Webserver
        node.purgediratrestart = config.purgediratrestart || "leave"; // 26/02/2020
        node.noderedport = typeof config.noderedport === "undefined" ? "1980" : config.noderedport;
        // 11/11/2019 NEW in V 1.1.0, changed webserver behaviour. Redirect pre V. 1.1.0 1880 ports to the nde default 1980
        if (node.noderedport.trim() == "1880") {
            RED.log.warn("SonosPollyTTS-config: The webserver port ist 1880. Please change it to another port, not to conflict with default node-red 1880 port. I've changed this temporarly for you to 1980");
            node.noderedport = "1980";
        }
        node.sNoderedURL = "http://" + node.noderedipaddress.trim() + ":" + node.noderedport.trim(); // 11/11/2019 New Endpoint to overcome https problem.
        RED.log.info('SonosPollyTTS-config: Node-Red node.js Endpoint will be created here: ' + node.sNoderedURL + "/tts");

        // 26/02/2020
        if (node.purgediratrestart === "purge") {
            // Delete all files, that are'nt OwnFiles_
            try {
                fs.readdirSync(path.join(node.userDir, "ttsfiles"), (err, files) => {
                    try {
                        if (files.length > 0) {
                            files.forEach(function (file) {
                                RED.log.info("SonospollyTTS-config: Deleted TTS file " + path.join(node.userDir, "ttsfiles", file));
                                try {
                                    fs.unlinkSync(path.join(node.userDir, "ttsfiles", file));
                                } catch (error) {
                                }
                            });
                        };
                    } catch (error) {

                    }

                });
            } catch (error) { }
        };




        // 11/11/2019 CREATE THE ENDPOINT
        // #################################
        const http = require('http')
        const sWebport = node.noderedport.trim();
        const requestHandler = (req, res) => {
            try {

                var url = require('url');
                var url_parts = url.parse(req.url, true);
                var query = url_parts.query;

                res.setHeader('Content-Disposition', 'attachment; filename=tts.mp3')
                if (fs.existsSync(query.f)) {
                    var readStream = fs.createReadStream(query.f);
                    readStream.on("error", function (err) {
                        fine(err);
                    });
                    readStream.pipe(res);
                    res.end;
                } else {
                    RED.log.error("SonosPollyTTS-config: Playsonos RED.httpAdmin file not found: " + query.f);
                    res.write("File not found");
                    res.end();
                }

            } catch (error) {
                RED.log.error("SonosPollyTTS-config: Playsonos RED.httpAdmin error: " + error + " on: " + query.f);
            }
            function fine(err) {
                RED.log.error("SonosPollyTTS-config: Playsonos error opening stream : " + query.f + ' : ' + error);
                res.end;
            }
        }


        try {
            node.oWebserver = http.createServer(requestHandler);
            node.oWebserver.on('error', function (e) {
                RED.log.error("SonosPollyTTS-config: " + node.ID + " error starting webserver on port " + sWebport + " " + e);
            });
        } catch (error) {
            // Already open. Close it and redo.
            RED.log.error("SonosPollyTTS-config: Webserver creation error: " + error);
        }

        try {
            node.oWebserver.listen(sWebport, (err) => {
                if (err) {
                    RED.log.error("SonosPollyTTS-config: error listening webserver on port " + sWebport + " " + err);
                }
            });

        } catch (error) {
            // In case oWebserver is null
            RED.log.error("SonosPollyTTS-config: error listening webserver on port " + sWebport + " " + error);
        }
        // #################################




        node.on('close', function (done) {
            // 11/11/2019 Close the Webserver
            try {
                node.oWebserver.close(function () { RED.log.info("SonosPollyTTS: Webserver UP. Closing down."); });
            } catch (error) {

            }
            setTimeout(function () {
                // Wait some time to allow time to do promises.
                done();
            }, 500);
        });


    }
    RED.nodes.registerType("sonospollytts-config", PollyConfigNode, {
        credentials: {
            accessKey: { type: "text" },
            secretKey: { type: "password" }
        }
    });

}

